src/helpers.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Synapse charm helpers."""
+
+
+import logging
+import typing
+
+import ops
+
+import pebble
+import signing_key
+import synapse
+from charm_state import CharmState
+
+logger = logging.getLogger(__name__)
+
+MAIN_UNIT_ID = 0
+
+
+def is_redis_required(charm: ops.CharmBase, charm_state: CharmState) -> bool:
+    """Check if Redis configuration should be required.
+
+    Args:
+        charm: charm instance.
+        charm_state: charm state.
+
+    Return:
+        True if more than 1 unit is found.
+    """
+    return charm_state.redis_config is None and planned_units(charm) > 1
+
+
+def is_mjolnir_enabled(charm: ops.CharmBase, charm_state: CharmState) -> bool:
+    """Check if Mjolnir should be enabled.
+
+    Args:
+        charm: charm instance.
+        charm_state: charm state.
+
+    Return:
+        True if is main and config is enabled.
+    """
+    return is_main(charm) and charm_state.synapse_config.enable_mjolnir
+
+
+def planned_units(charm: ops.CharmBase) -> int:
+    """Get number of planned units.
+
+    Args:
+        charm: charm instance.
+
+    Returns:
+        Number of planned units.
+    """
+    return charm.app.planned_units()
+
+
+def peer_relation(charm: ops.CharmBase) -> typing.Optional[ops.Relation]:
+    """Get peer relation.
+
+    Args:
+        charm: charm instance.
+
+    Returns:
+        Synapse peer relation.
+    """
+    peer_relations = charm.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
+    if not peer_relations:
+        return None
+    return peer_relations[0]
+
+
+def is_main(charm: ops.CharmBase) -> bool:
+    """Verify if this unit is the main.
+
+    Args:
+        charm: charm instance.
+
+    Returns:
+        bool: true if is the main unit.
+    """
+    return f"/{MAIN_UNIT_ID}" in charm.unit.name
+
+
+def get_unit_address(charm: ops.CharmBase, unit_id: int) -> str:
+    """Get unit address.
+
+    Args:
+        charm: charm instance.
+        unit_id: number as 0 in synapse/0.
+
+    Returns:
+        unit address as unit-0.synapse-endpoints.
+    """
+    return f"{charm.app.name}-{unit_id}.{charm.app.name}-endpoints"
+
+
+def get_unit_number(charm: ops.CharmBase) -> str:
```

Comment:
This function is only used once in the code. I know that it is a matter of preference but I wouldn't have written it. 
When appropriate, 
```
unit_number = charm.unit.name.split("/")[1]
```
does the job for me

Reply:
I agree. I just left it like this because I wanted to check with @dimaqq  if it would make sense to have this helper function in ops, since [other charms](https://github.com/search?q=int%28self.unit.name.split%28%22%2F%22%29%5B1%5D%29+org%3Acanonical&type=code) use it as well.